File: bench/suite/src/st/nbody/python/nbody.py
# n-body: single-threaded, pure CPython (Benchmarks Game algorithm).
import sys
from math import pi, sqrt

SOLAR_MASS = 4 * pi * pi
DAYS_PER_YEAR = 365.24


def body(x, y, z, vx, vy, vz, mass):
    return [x, y, z, vx * DAYS_PER_YEAR, vy * DAYS_PER_YEAR,
            vz * DAYS_PER_YEAR, mass * SOLAR_MASS]


BODIES = [
    body(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0),
    body(4.84143144246472090e+00, -1.16032004402742839e+00,
         -1.03622044471123109e-01, 1.66007664274403694e-03,
         7.69901118419740425e-03, -6.90460016972063023e-05,
         9.54791938424326609e-04),
    body(8.34336671824457987e+00, 4.12479856412430479e+00,
         -4.03523417114321381e-01, -2.76742510726862411e-03,
         4.99852801234917238e-03, 2.30417297573763929e-05,
         2.85885980666130812e-04),
    body(1.28943695621391310e+01, -1.51111514016986312e+01,
         -2.23307578892655734e-01, 2.96460137564761618e-03,
         2.37847173959480950e-03, -2.96589568540237556e-05,
         4.36624404335156298e-05),
    body(1.53796971148509165e+01, -2.59193146099879641e+01,
         1.79258772950371181e-01, 2.68067772490389322e-03,
         1.62824170038242295e-03, -9.51592254519715870e-05,
         5.15138902046611451e-05),
]

PAIRS = [(BODIES[i], BODIES[j])
         for i in range(len(BODIES)) for j in range(i + 1, len(BODIES))]


def offset_momentum():
    px = py = pz = 0.0
    for b in BODIES:
        px += b[3] * b[6]
        py += b[4] * b[6]
        pz += b[5] * b[6]
    sun = BODIES[0]
    sun[3] = -px / SOLAR_MASS
    sun[4] = -py / SOLAR_MASS
    sun[5] = -pz / SOLAR_MASS


def advance(dt, steps):
    pairs = PAIRS
    bodies = BODIES
    for _ in range(steps):
        for bi, bj in pairs:
            dx = bi[0] - bj[0]
            dy = bi[1] - bj[1]
            dz = bi[2] - bj[2]
            d2 = dx * dx + dy * dy + dz * dz
            mag = dt / (d2 * sqrt(d2))
            mj = bj[6] * mag
            mi = bi[6] * mag
            bi[3] -= dx * mj
            bi[4] -= dy * mj
            bi[5] -= dz * mj
            bj[3] += dx * mi
            bj[4] += dy * mi
            bj[5] += dz * mi
        for b in bodies:
            b[0] += dt * b[3]
            b[1] += dt * b[4]
            b[2] += dt * b[5]


def energy():
    e = 0.0
    for b in BODIES:
        e += 0.5 * b[6] * (b[3] * b[3] + b[4] * b[4] + b[5] * b[5])
    for bi, bj in PAIRS:
        dx = bi[0] - bj[0]
        dy = bi[1] - bj[1]
        dz = bi[2] - bj[2]
        e -= (bi[6] * bj[6]) / sqrt(dx * dx + dy * dy + dz * dz)
    return e


def main():
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 1000
    offset_momentum()
    print("%.9f" % energy())
    advance(0.01, n)
    print("%.9f" % energy())


main()

File: bench/suite/src/st/nbody/rust/nbody.rs
// n-body: single-threaded, same algorithm as the Benchmarks Game description.
use std::f64::consts::PI;

const SOLAR_MASS: f64 = 4.0 * PI * PI;
const DAYS_PER_YEAR: f64 = 365.24;

#[derive(Clone, Copy)]
struct Body {
    x: f64,
    y: f64,
    z: f64,
    vx: f64,
    vy: f64,
    vz: f64,
    mass: f64,
}

fn bodies() -> [Body; 5] {
    [
        // Sun
        Body { x: 0.0, y: 0.0, z: 0.0, vx: 0.0, vy: 0.0, vz: 0.0, mass: SOLAR_MASS },
        // Jupiter
        Body {
            x: 4.84143144246472090e+00,
            y: -1.16032004402742839e+00,
            z: -1.03622044471123109e-01,
            vx: 1.66007664274403694e-03 * DAYS_PER_YEAR,
            vy: 7.69901118419740425e-03 * DAYS_PER_YEAR,
            vz: -6.90460016972063023e-05 * DAYS_PER_YEAR,
            mass: 9.54791938424326609e-04 * SOLAR_MASS,
        },
        // Saturn
        Body {
            x: 8.34336671824457987e+00,
            y: 4.12479856412430479e+00,
            z: -4.03523417114321381e-01,
            vx: -2.76742510726862411e-03 * DAYS_PER_YEAR,
            vy: 4.99852801234917238e-03 * DAYS_PER_YEAR,
            vz: 2.30417297573763929e-05 * DAYS_PER_YEAR,
            mass: 2.85885980666130812e-04 * SOLAR_MASS,
        },
        // Uranus
        Body {
            x: 1.28943695621391310e+01,
            y: -1.51111514016986312e+01,
            z: -2.23307578892655734e-01,
            vx: 2.96460137564761618e-03 * DAYS_PER_YEAR,
            vy: 2.37847173959480950e-03 * DAYS_PER_YEAR,
            vz: -2.96589568540237556e-05 * DAYS_PER_YEAR,
            mass: 4.36624404335156298e-05 * SOLAR_MASS,
        },
        // Neptune
        Body {
            x: 1.53796971148509165e+01,
            y: -2.59193146099879641e+01,
            z: 1.79258772950371181e-01,
            vx: 2.68067772490389322e-03 * DAYS_PER_YEAR,
            vy: 1.62824170038242295e-03 * DAYS_PER_YEAR,
            vz: -9.51592254519715870e-05 * DAYS_PER_YEAR,
            mass: 5.15138902046611451e-05 * SOLAR_MASS,
        },
    ]
}

fn offset_momentum(b: &mut [Body; 5]) {
    let (mut px, mut py, mut pz) = (0.0, 0.0, 0.0);
    for bi in b.iter() {
        px += bi.vx * bi.mass;
        py += bi.vy * bi.mass;
        pz += bi.vz * bi.mass;
    }
    b[0].vx = -px / SOLAR_MASS;
    b[0].vy = -py / SOLAR_MASS;
    b[0].vz = -pz / SOLAR_MASS;
}

fn energy(b: &[Body; 5]) -> f64 {
    let mut e = 0.0;
    for i in 0..b.len() {
        let bi = &b[i];
        e += 0.5 * bi.mass * (bi.vx * bi.vx + bi.vy * bi.vy + bi.vz * bi.vz);
        for bj in &b[i + 1..] {
            let dx = bi.x - bj.x;
            let dy = bi.y - bj.y;
            let dz = bi.z - bj.z;
            e -= (bi.mass * bj.mass) / (dx * dx + dy * dy + dz * dz).sqrt();
        }
    }
    e
}

fn advance(b: &mut [Body; 5], dt: f64) {
    for i in 0..b.len() {
        let (left, right) = b.split_at_mut(i + 1);
        let bi = &mut left[i];
        for bj in right.iter_mut() {
            let dx = bi.x - bj.x;
            let dy = bi.y - bj.y;
            let dz = bi.z - bj.z;
            let d2 = dx * dx + dy * dy + dz * dz;
            let mag = dt / (d2 * d2.sqrt());
            let mj = bj.mass * mag;
            bi.vx -= dx * mj;
            bi.vy -= dy * mj;
            bi.vz -= dz * mj;
            let mi = bi.mass * mag;
            bj.vx += dx * mi;
            bj.vy += dy * mi;
            bj.vz += dz * mi;
        }
    }
    for bi in b.iter_mut() {
        bi.x += dt * bi.vx;
        bi.y += dt * bi.vy;
        bi.z += dt * bi.vz;
    }
}

fn main() {
    let n: usize = std::env::args().nth(1).and_then(|s| s.parse().ok()).unwrap_or(1000);
    let mut b = bodies();
    offset_momentum(&mut b);
    println!("{:.9}", energy(&b));
    for _ in 0..n {
        advance(&mut b, 0.01);
    }
    println!("{:.9}", energy(&b));
}
